### DISCUSSION TITLE
Any contribution tutorial?

### DISCUSSION THREAD

**Discussion by caibirdme** [vectordotdev/vector#12762] [Q&A] [answered]
I want to extend vector to support source like pulsar and sinks that suitable for our infrustructures. But I don't know how to get started.
I found devloping docs, but that just tell me how to setup dev env, and the arch doc is just a very simple description of how components are connected. And nothing more!

It'd be better if there's a documentation that defines what a (source/sink) plugin is (the interface definition in java or go, or trait in rust), if people impl that trait how to tell the core to identify it, and this can help people make sense.

further more, the doc may contain an simple example of how to develop your own sources\sinks (maybe a simple http source with very few configs and ignore some complicated err handlings)
    * where to add the code
    * where to define the config strcuture
    * how to setup a server listen on configured port
    * how to set decoder to decode data received from that connection,
    * how to send data to the next pipeline

**Comment by jszwedko** [comment, marked as answer]
Hi @caibirdme !

I agree that we could document how to add new components better; this is something we've been thinking about, including creating more clear boundaries between them. With the current state of things, contributors usually will copy an existing component that is close to what they want to do and then just modify that. For the `pulsar` sink, copying the `kafka` sink is probably a reasonable place to start. Happy to give more feedback on a draft PR if you do want to take a stab at it!

**Reply by caibirdme** [reply]
I'm trying to make a opentelemetry-log source plugin and I met with some problems.
1. I compiled the code and run `./vector list`, it shows the plugin name I've made
2. But when I run `./vector -c xx.yaml` it reports an error: `Configuration error. error=unknown variant "otel_log", expected one of xxx`
3. Am I mising something?
4. When I run `cargo build`, it warned me that my fn is unused, so do I need update somewhere else to make the plugin available?
5. I added the code, but it still doesn't work
```inventory::submit! {
    SourceDescription::new::<OpentelemetryLogConfig>("otel_log")
}
```

**Reply by jszwedko** [reply]
Hi @caibirdme ! If you wanted to push up a branch, we could take a look. It should be available if your configuration struct implements `SourceConfig` like you see here:

https://github.com/vectordotdev/vector/blob/8e640f1a867d92e313461f18178e375d1d4c4dad/src/sources/kafka.rs#L137-L139

**Reply by spencergilbert** [reply]
Might also need to be added to the features to actually be build when Vector is compiled

**Reply by caibirdme** [reply]
> Hi @caibirdme ! If you wanted to push up a branch, we could take a look. It should be available if your configuration struct implements `SourceConfig` like you see here:
> 
> https://github.com/vectordotdev/vector/blob/8e640f1a867d92e313461f18178e375d1d4c4dad/src/sources/kafka.rs#L137-L139

Thx, I missed the `#[typetag::serde(name = "kafka")]` and now it works!
